### ISSUE TITLE
[RFC] Move `eval_metric` in `XGBModel.fit` to `__init__` and accept sklearn style metrics.

### ISSUE THREAD

**Issue by trivialfis** [dmlc/xgboost#6735]
This is a RFC for Python sklearn interface.  Currently we support custom evaluation metric via `eval_metric` parameter from `XGBModel.fit` method.  The accepted values are string or list of strings for builtin metrics, native custom evaluation metric that accepts a `DMatrix` for user defined metrics.  This RFC is to suggest 2 related changes:

- Replace it with a parameter in `__init__` method.
- Accept `sklearn.metrics` style metrics.

## Pros
- This helps aligning xgboost with sklearn estimator design guide: https://scikit-learn.org/stable/developers/develop.html#fitting  , which states that parameter in `fit` method should be strictly data dependent, and `eval_metric` doesn't seem to meet such requirement.
- This helps users who are familiar with sklearn metrics to avoid defining a wrapper that accept `DMatrix`.
- This makes custom evaluation metric consistent with custom objective function in sklearn interface, where xgboost defines a `_objective_decorator` internally to wrap around sklearn style functions.

If we make the changes, user can simply pass any metric in `sklearn.metrics` into `XGBModel` and `DaskXGBModel`.

## Cons
- It will be a breaking change.  But since we are replacing it, so both parameters can coexist for a few releases.

## Implementation

We can define a similar decorator like `_objective_decorator` for metric:

``` python
def _metric_decorator(func: Callable) -> Metric:
    # func is a metric like `sklearn.metrics.roc_auc_score`.
    def inner(y_score: np.ndarray, dmatrix: DMatrix) -> float:
        y_true = dmatrix.get_label()
        return func.__name__, func(y_true, y_score)
    return inner
```

**Comment by hcho3** [comment]
I think this proposal is consistent with our goal of first-class integration with scikit-learn, and it has relatively low risk.

**Comment by RAMitchell** [comment]
Sounds good to me.

**Comment by trivialfis** [comment]
I will target this for 1.5.

**Comment by trivialfis** [comment]
https://discuss.xgboost.ai/t/nan-values-with-early-stopping/2494/3

**Comment by trivialfis** [comment]
Reproducible example: https://github.com/mangleddata/xgbnan .
